### ISSUE TITLE
LoRA training on krea2 grows rank-1 spectral spikes, breaking multi-LoRA stacking (measurements + mitigation)

### ISSUE THREAD

**Issue by CoreyCorza** [ostris/ai-toolkit#964]
## This is for bugs only

Did you already ask [in the discord](https://discord.gg/VXmU2f5WEU)?

No

You verified that this is a bug and not a feature request or question by asking [in the discord](https://discord.gg/VXmU2f5WEU)?

No

## Corey's Summary
Using multiple LoRA's in comfyui, trained with ai-toolkit, always seem to fight each other and degrade image quality, no matter how the LoRA is trained. LoKr being the exception.
Got Fable to run tests on known high quality checkpoints vs poor vs LoKr.
Fable reproduced the issue with a temporary working solution, that seemingly eliminates all the multi-lora conflictions on inference of which I have verified as being consistent.
Fable would like to share the results of its findings.

## Fable's Summary

LoRA training on the krea2 arch appears to grow an **unbounded, effectively rank-1 spectral spike** in the per-module weight deltas, concentrated on the SwiGLU `mlp.gate` modules (and `txtfusion.*.mlp.down`). The spike grows monotonically with training steps and occurs with **both automagic3 and adamw8bit** (automagic3 just reaches it ~10x faster). Full-rank LoKr trained in the same tool does **not** exhibit it.

Single-LoRA behavior looks fine but any two such LoRAs loaded together in ComfyUI produce severe quality collapse. Post-hoc spectral clipping of the checkpoints (cap per-module sigma_max) restores clean stacking, which confirms the spike is the cause rather than content overlap.

## Measurements

Per-module singular values of the effective delta (`B@A`, alpha-scaled; for LoKr, `kron(w1,w2)` via norm factorization). "Stable rank" = ||d||_F^2 / sigma_max^2 (≈ how many directions the update energy is spread over).

| checkpoint | type | optimizer / timestep | steps | peak sigma_max | stable rank of top modules |
|---|---|---|---|---|---|
| character A, r64 | LoRA | automagic3 / sigmoid | 2,100 | **20.3** | ~1.0 |
| character B, r32 | LoRA | automagic3 / sigmoid | 3,000 | **23.1** | ~1.0 |
| concept, r64 | LoRA | adamw8bit / linear, lr 7.5e-5 | 16,400 | 14.0 | ~1.0 |
| same run, later | LoRA | adamw8bit / linear | 26,000 | 22.4 | ~1.0 |
| concept, r128 | LoRA | adamw8bit / linear, lr 5e-5 | 60,000 | **27.6** | ~1.0 |
| concept | LoKr full-rank, factor 8 | (same tool) | ~8,700 | **2.9** | mean stable rank ≈ 115 |
| third-party civitai krea2 LoKr (stacks fine with other LoRA's) | LoKr full-rank, factor 4 | unknown | unknown | **4.5** | mean stable rank ≈ 86 |

Key observations:

- The spike **grows with steps** (14.0 → 22.4 → 27.6 across one adamw8bit run's checkpoints), so it is not an optimizer bug per se — it looks like the implicit bias of the low-rank `B@A` parameterization on this arch, with automagic3 acting as an accelerant (σ ~20 in ~2k steps vs ~26k for adamw).
- The top-magnitude modules are consistently `blocks.*.mlp.gate` (SwiGLU gate — multiplicative, so most saturation-sensitive) and `txtfusion.layerwise_blocks.*.mlp.down`.
- Rank (8/16/32/64/128), lr, max_lr caps, quantization (none/float8/8bit convrot), diff-output-preservation, and timestep type made **no measurable difference** to the collapse — everything converges to stable rank ≈ 1 with large sigma.
- Cross-LoRA subspace alignment between two affected LoRAs is ~0 (mean cosine 0.006), i.e. the stacking failure is **not** interference between the LoRAs' content — it's both of them independently pushing gate activations off-distribution.

## Confirmation experiment

SVD each module's delta, cap singular values at 5.0 (chosen to match the healthy LoKr's ceiling), re-factor into `B@A`, save. Result (user-blind A/B in ComfyUI, two LoRAs stacked, same seeds): unclipped pair = collapsed anatomical detail; clipped pair = clean output with both LoRAs functioning. Night-and-day difference, reproduced across two independent LoRA pairs (one automagic3-trained, one adamw8bit-trained).

## Reproduction script

Point at any krea2 LoRA safetensors (pure Python + numpy, no torch needed):

<details>
<summary>spectra.py — measure per-module sigma_max / stable rank</summary>

```python
import json, struct, sys
from collections import defaultdict
import numpy as np

def read_header(path):
    with open(path, "rb") as f:
        (hlen,) = struct.unpack("<Q", f.read(8))
        h = json.loads(f.read(hlen))
    h.pop("__metadata__", None)
    return h, 8 + hlen

def load(path, info, base):
    start, end = info["data_offsets"]
    with open(path, "rb") as f:
        f.seek(base + start); raw = f.read(end - start)
    if info["dtype"] == "BF16":
        u = np.frombuffer(raw, dtype=np.uint16).astype(np.uint32) << 16
        a = u.view(np.float32).astype(np.float32)
    else:
        a = np.frombuffer(raw, dtype={"F32":np.float32,"F16":np.float16}[info["dtype"]]).astype(np.float32)
    return a.reshape(info["shape"])

def analyze(path):
    h, base = read_header(path)
    lokr = defaultdict(dict); lora = defaultdict(dict)
    for k in h:
        if k.endswith(".lokr_w1"): lokr[k[:-8]]["w1"] = k
        elif k.endswith(".lokr_w2"): lokr[k[:-8]]["w2"] = k
        elif k.endswith(".lora_A.weight"): lora[k[:-len(".lora_A.weight")]]["A"] = k
        elif k.endswith(".lora_B.weight"): lora[k[:-len(".lora_B.weight")]]["B"] = k
    stats = []
    for mod, parts in sorted(lokr.items()):
        if "w1" not in parts or "w2" not in parts: continue
        w1 = load(path, h[parts["w1"]], base); w2 = load(path, h[parts["w2"]], base)
        s1 = np.linalg.svd(w1, compute_uv=False); s2 = np.linalg.svd(w2, compute_uv=False)
        smax = float(s1[0] * s2[0])
        fro2 = float((s1**2).sum() * (s2**2).sum())
        stats.append((mod, smax, fro2 / (smax**2)))
    for mod, parts in sorted(lora.items()):
        if "A" not in parts or "B" not in parts: continue
        A = load(path, h[parts["A"]], base); B = load(path, h[parts["B"]], base)
        if A.ndim > 2: A = A.reshape(A.shape[0], -1)
        if B.ndim > 2: B = B.reshape(B.shape[0], -1)
        Qb, Rb = np.linalg.qr(B); Qa, Ra = np.linalg.qr(A.T)
        s = np.linalg.svd(Rb @ Ra.T, compute_uv=False)
        smax = float(s[0]); fro2 = float((s**2).sum())
        stats.append((mod, smax, fro2 / (smax**2)))
    smaxes = np.array([x[1] for x in stats]); sranks = np.array([x[2] for x in stats])
    print(f"\n{path.split('/')[-1]}  ({len(stats)} modules)")
    print(f"  sigma_max   : mean={smaxes.mean():8.4f}  median={np.median(smaxes):8.4f}  max={smaxes.max():8.4f}")
    print(f"  stable rank : mean={sranks.mean():8.2f}  median={np.median(sranks):8.2f}  min={sranks.min():8.2f}")
    stats.sort(key=lambda x: -x[1])
    for mod, smax, sr in stats[:6]:
        print(f"    smax={smax:8.4f}  stable_rank={sr:7.2f}  {mod}")

for p in sys.argv[1:]:
    analyze(p)
```

</details>

<details>
<summary>clip_sigma.py — mitigation: cap per-module sigma and re-save</summary>

```python
import json, struct, sys
from collections import defaultdict
import numpy as np

CAP = float(sys.argv[2]) if len(sys.argv) > 2 else 5.0

def read_header(path):
    with open(path, "rb") as f:
        (hlen,) = struct.unpack("<Q", f.read(8))
        h = json.loads(f.read(hlen))
    h.pop("__metadata__", None)
    return h, 8 + hlen

def load(path, info, base):
    start, end = info["data_offsets"]
    with open(path, "rb") as f:
        f.seek(base + start); raw = f.read(end - start)
    if info["dtype"] == "BF16":
        u = np.frombuffer(raw, dtype=np.uint16).astype(np.uint32) << 16
        a = u.view(np.float32).astype(np.float32)
    else:
        a = np.frombuffer(raw, dtype={"F32":np.float32,"F16":np.float16}[info["dtype"]]).astype(np.float32)
    return a.reshape(info["shape"])

def to_bf16_bytes(arr):
    f = np.ascontiguousarray(arr, dtype=np.float32)
    u = f.view(np.uint32)
    rounded = ((u + 0x7FFF + ((u >> 16) & 1)) >> 16).astype(np.uint16)
    nan_mask = np.isnan(f)
    if nan_mask.any():
        rounded[nan_mask] = 0x7FC0
    return rounded.tobytes()

src = sys.argv[1]
h, base = read_header(src)
mods = defaultdict(dict)
for k in h:
    if k.endswith(".lora_A.weight"): mods[k[:-len(".lora_A.weight")]]["A"] = k
    elif k.endswith(".lora_B.weight"): mods[k[:-len(".lora_B.weight")]]["B"] = k

out_tensors = {}
clipped = 0
for mod, parts in sorted(mods.items()):
    A = load(src, h[parts["A"]], base); B = load(src, h[parts["B"]], base)
    a_shape, b_shape = A.shape, B.shape
    A2 = A.reshape(A.shape[0], -1) if A.ndim > 2 else A
    B2 = B.reshape(B.shape[0], -1) if B.ndim > 2 else B
    Qb, Rb = np.linalg.qr(B2); Qa, Ra = np.linalg.qr(A2.T)
    U, s, Vt = np.linalg.svd(Rb @ Ra.T)
    if s[0] > CAP: clipped += 1
    root = np.sqrt(np.minimum(s, CAP))
    newB = (Qb @ U) * root[None, :]
    newA = (root[:, None] * Vt) @ Qa.T
    out_tensors[parts["B"]] = newB.reshape(b_shape)
    out_tensors[parts["A"]] = newA.reshape(a_shape)

dst = src.replace(".safetensors", f"_sclip{CAP:g}.safetensors")
header = {}; offset = 0; blobs = []
for k in out_tensors:
    b = to_bf16_bytes(out_tensors[k])
    header[k] = {"dtype": "BF16", "shape": list(out_tensors[k].shape), "data_offsets": [offset, offset + len(b)]}
    blobs.append(b); offset += len(b)
hj = json.dumps(header).encode()
with open(dst, "wb") as f:
    f.write(struct.pack("<Q", len(hj))); f.write(hj)
    for b in blobs: f.write(b)
print(f"wrote {dst} | modules clipped: {clipped}/{len(mods)} (cap={CAP})")
```

</details>

## Environment

- ai-toolkit latest `main` as of 2026-07-19 (also reproduced on builds from ~2 weeks prior), torch/triton current
- Base model: `krea/Krea-2-Raw`, arch `krea2`, bf16, various quantization settings (issue reproduces with quantization disabled)
- Inference: ComfyUI, stock LoRA loading, fresh install reproduced it

## Caveats

- One user's datasets (N≈5 runs measured), one arch — I have not verified whether other DiT arches in the toolkit show the same growth.
- The cap value 5.0 was chosen empirically to match healthy LoKr ceilings, not tuned rigorously.
- I'm not claiming clipping is the right in-trainer fix — possible directions might be spectral-norm regularization, clip-on-save as an option, or documenting full-rank LoKr as the recommended network type for krea2.
